lib/services/data/providers/loki.js

```diff
@@ -0,0 +1,459 @@
+    }
+
+    if (options.fields) {
+      returnArray = returnArray.map(item => {
+        return _.pick(item, Object.keys(options.fields));
+      });
+    }
+    return returnArray;
+  }
+  find(path, parameters, callback) {
+    if (typeof parameters === 'function') {
+      callback = parameters;
+      parameters = {};
+    }
+    if (parameters == null) {
+      parameters = {};
+    }
+    let returnArray = [];
+    try {
+      returnArray = this.findInternal(path, parameters);
+    } catch (e) {
+      callback(e);
+      return;
+    }
+    callback(null, returnArray);
+  }
+  count(path, parameters, callback) {
+    let pathCriteria = this.getPathCriteria(path);
+    if (parameters.criteria) {
+      pathCriteria = this.addCriteria(this.getPathCriteria(path), parameters.criteria);
+    }
+    let count = 0;
+    try {
+      count = this.collection.count(pathCriteria);
+    } catch (e) {
+      callback(e);
+      return;
+    }
+    callback(null, count);
+  }
+  findOne(criteria, fields, callback) {
+    if (typeof fields === 'function') {
+      callback = fields;
+      fields = null;
+    }
+    let result = null;
+    try {
+      result = this.findOneInternal(criteria, fields);
+    } catch (e) {
+      callback(e);
+      return;
+    }
+    callback(null, result);
+  }
+  findOneInternal(criteria, fields) {
+    let results = this.findInternal('*', { criteria, fields });
+    if (results.length > 0) {
+      return results[0];
+    }
+    return null;
+  }
+  addCriteria(pathObject, criteria) {
+    return { $and: [pathObject, criteria] };
+  }
+  getPathCriteria(path) {
+    if (path.indexOf('*') === -1) {
+      return { path };
+    }
+    return {
+      path: { $regex: '^' + this.escapeRegex(this.preparePath(path)).replace(/\\\*/g, '.*') + '$' }
+    };
+  }
+  preparePath(path) {
+    return path.replace(/(\*)\1+/g, '$1');
+  }
+  escapeRegex(str) {
+    return str.replace(/[|\\{}()[\]^$+*?.]/g, '\\$&');
+  }
+  increment(path, counterName, increment, callback) {
+    this.operationQueue.push(
+      {
+        operationType: constants.DATA_OPERATION_TYPES.INCREMENT,
+        arguments: [path, counterName, increment]
+      },
+      callback
+    );
+  }
+  incrementInternal(path, counterName, increment) {
+    let recordToIncrement = this.findOneInternal({ path }) || {
+      data: { [counterName]: { value: 0 } }
+    };
+    if (recordToIncrement.data[counterName] == null) {
+      recordToIncrement.data[counterName] = { value: 0 };
+    }
+    recordToIncrement.data[counterName].value += increment;
+    this.upsertInternal(path, recordToIncrement);
+    return recordToIncrement.data[counterName].value;
+  }
+  getMeta(document) {
+    return {
+      created: document.created,
+      modified: document.modified,
+      modifiedBy: document.modifiedBy,
+      path: document.path
+    };
+  }
+  transform(dataObj, meta) {
+    return {
+      data: dataObj.data,
+      _meta: meta || {
```

Comment:
why are we not storing metadata under the _meta path in dataObj? That would offer better protection against name collisions

Reply:
is this determined by lokijs?